src/fs/file.rs

```diff
@@ -40,12 +42,16 @@ pub struct File<'dir> {
     /// path (following a symlink).
     pub path: PathBuf,
 
+    /// The type of this file. None for directories `.` and `..`.
+    file_type: Option<FileType>,
+
     /// A cached `metadata` (`stat`) call for this file.
     ///
     /// This too is queried multiple times, and is *not* cached by the OS, as
     /// it could easily change between invocations — but exa is so short-lived
     /// it’s better to just cache it.
-    pub metadata: std::fs::Metadata,
+    /// It may be None if the call to symlink_metadata fails.
+    metadata: OnceCell<Option<Metadata>>,
```

Comment:
I’m not sure why you’re using `OnceCell` here (to be honest I’ve checked it and didn’t really understood what it was about), could you explain why you used it?

Reply:
In case of colorful output, we will determine if a file `is_executable_file`, which requires metadata (for permission informatino). But if not, we don't need `Metadata` at all. So I made it a lazy field: only invoke syscall if really needed.
We construct `File`s during scanning, not rendering. So it is not easy to determine whether this field is needed when constructing `File`.

BTW: `unsync::OnceCell` should be more efficient. But we use multi-threading and pass `&File` to other threads. Therefore, only `sync::OnceCell` can be used, though there should be only one thread accessing it at a time.

Reply:
Well, if I understand correctly, the only case where `metadata` is not initialized should be with `.` and `..`. Could the usage of `OnceCell` be avoided by checking `is_all_all` in the `metadata()` function?

Reply:
> the only case where metadata is not initialized should be with . and ..

No. In my benchmark in the PR description (used as `exa /nix/store | wc -l`), there is no `metadata` being called at all because color is disabled for pipe. That's why it's 10x faster.
`.` and `..` are already not calling `metadata` on master. That's not the key point.